### PR TITLE
KAFKA-15177: Implement KIP-875 SourceConnector::alterOffset API in MirrorMaker 2 connectors

### PR DESCRIPTION
**Pull request by C0urante**
[Jira](https://issues.apache.org/jira/browse/KAFKA-15177), [KIP-875](https://cwiki.apache.org/confluence/display/KAFKA/KIP-875%3A+First-class+offsets+support+in+Kafka+Connect)

### Summary
Implements the new `alterOffsets` method in all three MM2 connectors. The only connector that makes use of the Connect framework-managed offsets at runtime is the `MirrorSourceConnector`, but since the other two connectors also emit offsets with their records, we also allow users to add/remove offsets for them as well, as long as the offsets match the format of the ones that are emitted by the connectors.

### Integration testing
An integration test case is added (or really, augmented) that focuses primarily on the `MirrorSourceConnector`. This case piggybacks off of the existing `MirrorConnectorsExactlyOnceIntegrationTest::testReplication` test case, which is done for a few reasons:
- Every new case added to the `MirrorConnectorsIntegrationBaseTest` is run by that class and all of its subclasses (of which there are currently five), which can lead to bloat in testing runtime
- Graceful shutdown of MM2 tasks may not always occur in a timely fashion, which can cause them to emit stale offsets after they are supposed to have been stopped; running with exactly-once support eliminates this risk
- Some subclasses of the `MirrorConnectorsIntegrationBaseTest` (specifically, the `MirrorConnectorsIntegrationTransactionsTest` suite) cause unpredictable behavior with offsets that can make it harder to verify an exact number of records in a replicated topic due to, e.g., control records

Still, this is a little hacky. Happy to change things up if we think it's worth the work to find something cleaner.

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

### CODE REVIEW

File: connect/runtime/src/test/java/org/apache/kafka/connect/util/clusters/EmbeddedKafkaCluster.java [hunk 1]
```diff
@@ -597,7 +596,9 @@ private Set<TopicPartition> listPartitions(
             Admin admin,
             Collection<String> topics
     ) throws TimeoutException, InterruptedException, ExecutionException {
-        assertFalse("collection of topics may not be empty", topics.isEmpty());
```
Comment (C0urante) [inline connect/runtime/src/test/java/org/apache/kafka/connect/util/clusters/EmbeddedKafkaCluster.java:600]: Removed assertions from this API as it added an implicit dependency on JUnit 4, which we don't use in the `:connect:mirror` module.

**Comment by C0urante** [comment]
@yashmayya do you have time to take a look at this? Feel like it may be useful to you with the work we were discussing on https://github.com/apache/kafka/pull/13945 about how we want to set examples with our implementation of this API, and possibly modifying the types of the objects that the framework passes to the `alterOffsets` method.

File: connect/mirror/src/main/java/org/apache/kafka/connect/mirror/MirrorCheckpointConnector.java [hunk 2]
```diff
@@ -132,6 +136,27 @@ public String version() {
         return AppInfoParser.getVersion();
     }
 
+    @Override
+    public boolean alterOffsets(Map<String, String> connectorConfig, Map<Map<String, ?>, Map<String, ?>> offsets) {
```
Comment (yashmayya) [inline connect/mirror/src/main/java/org/apache/kafka/connect/mirror/MirrorCheckpointConnector.java:140]: What's the benefit of allowing users to modify the offsets here? It seems to me like it's essentially a no-op, so would it not be more appropriate to simply throw an `UnsupportedOperationException` with a relevant message?

File: connect/mirror/src/main/java/org/apache/kafka/connect/mirror/MirrorHeartbeatConnector.java [hunk 3]
```diff
@@ -85,6 +89,26 @@ public String version() {
         return AppInfoParser.getVersion();
     }
 
+    @Override
+    public boolean alterOffsets(Map<String, String> config, Map<Map<String, ?>, Map<String, ?>> offsets) {
```
Comment (yashmayya) [inline connect/mirror/src/main/java/org/apache/kafka/connect/mirror/MirrorHeartbeatConnector.java:93]: Same question as the `MirrorCheckpointConnector` case - why do we want to allow users to modify the offsets?

File: connect/mirror/src/test/java/org/apache/kafka/connect/mirror/integration/MirrorConnectorsIntegrationExactlyOnceTest.java [hunk 4]
```diff
@@ -46,4 +51,49 @@ public void startClusters() throws Exception {
...
+                "Records were not re-replicated to backup cluster after altering offsets.");
+        int expectedRecordsTopic2 = NUM_RECORDS_PER_PARTITION * 2;
+        assertEquals(expectedRecordsTopic2, backup.kafka().consume(expectedRecordsTopic2, RECORD_TRANSFER_DURATION_MS, backupTopic2).count(),
+                "New topic was not re-replicated to backup cluster after altering offsets.");
+
+        @SuppressWarnings({"unchecked", "rawtypes"})
+        Class<? extends Connector>[] connectorsToReset = CONNECTOR_LIST.toArray(new Class[0]);
+        // Resetting the offsets for the heartbeat and checkpoint connectors doesn't have any effect
+        // on their behavior, but users may want to wipe offsets from them to prevent the offsets topic
+        // from growing infinitely. So, we include them in the list of connectors to reset as a sanity check
```
Comment (yashmayya) [inline connect/mirror/src/test/java/org/apache/kafka/connect/mirror/integration/MirrorConnectorsIntegrationExactlyOnceTest.java:88]: Is this also the reason for not just throwing an `UnsupportedOperationException` in their `alterOffsets` methods? If so, we don't really need validation logic there right?

**Review by yashmayya — COMMENTED** [review]
Thanks Chris. I'm not too familiar with MirrorMaker but the changes here seem reasonable to me. I only had a couple of questions around the behavior for `MirrorCheckpointConnector` and `MirrorHeartbeatConnector`.

File: connect/mirror/src/test/java/org/apache/kafka/connect/mirror/integration/MirrorConnectorsIntegrationExactlyOnceTest.java (hunk 4 above)
Comment (C0urante) [reply connect/mirror/src/test/java/org/apache/kafka/connect/mirror/integration/MirrorConnectorsIntegrationExactlyOnceTest.java:88]: Heh, I was wondering if we'd have to do a deep dive on the rationale here.

There are a few thoughts I had about what kind of offset tweaking to allow for the checkpoint and heartbeat connectors:
- (As noted in the comment here) Users may want to tombstone offsets for these connectors to prevent the offsets topic from growing infinitely
- We have no way of knowing if the set of offsets given to us is the total set or not, so we can't choose to only allow total resets instead of partial resets
- Since partial resets become possible, it seems reasonable to allow users to "undo" a partial reset by re-submitting an offset for a given partition
- This is especially relevant since, although the offsets topic isn't public API, its contents are now public API (via the `GET /connectors/{name}/offsets` endpoint), and users may want to track the offsets for these connectors for monitoring purposes
- We may make use of these offsets in future versions of these connectors, at which point, it'll be nice to know that the only offsets that we can ever encounter are the ones that were either emitted by the connector, or at least match the format of ones emitted by the connector

Also, I think the `UnsupportedOperationException` case was more intended for when offsets are managed externally, not for when it doesn't seem to make sense to modify them (because, e.g., they're never read back by the connector or its tasks).

So with all this in mind, I figured it'd be best to allow offsets to be modified by users, but only if they match the format of the offsets that the connector's tasks already emit. Thoughts?

As an aside... I'm wondering if we should abandon all kinds of validation when the proposed partition/offset pair is a tombstone. Just in case the offsets topic somehow ends up with garbage in it, it'd be nice to allow users to clean it up via the REST API, and if there's no existing partition/offset pair in the offsets topic for the given partition already, then emitting a tombstone effectively becomes a no-op. Thoughts x2?

Comment (yashmayya) [reply connect/mirror/src/test/java/org/apache/kafka/connect/mirror/integration/MirrorConnectorsIntegrationExactlyOnceTest.java:88]: > (As noted in the comment here) Users may want to tombstone offsets for these connectors to prevent the offsets topic from growing infinitely

Since the set of source partitions is limited here, shouldn't log compaction be good enough?

> We have no way of knowing if the set of offsets given to us is the total set or not, so we can't choose to only allow total resets instead of partial resets

Yeah, that makes sense.

> Since partial resets become possible, it seems reasonable to allow users to "undo" a partial reset by re-submitting an offset for a given partition

If these offsets are never read back and actually used, why would users want to "undo" partial or complete resets?

> Also, I think the UnsupportedOperationException case was more intended for when offsets are managed externally, not for when it doesn't seem to make sense to modify them (because, e.g., they're never read back by the connector or its tasks).
>
> So with all this in mind, I figured it'd be best to allow offsets to be modified by users, but only if they match the format of the offsets that the connector's tasks already emit. Thoughts?

Hm, I still feel like if offset modification isn't really doing anything here it might be more intuitive to simply disallow it rather than informing users that the offsets have been modified successfully but with no actual side-effect. However, I don't think too many users are going to be attempting to modify the offsets for `MirrorCheckpointConnector` / `MirrorHeartbeatConnector` and expecting any change in their behavior so I'm not too fussed about this either way.

> As an aside... I'm wondering if we should abandon all kinds of validation when the proposed partition/offset pair is a tombstone. Just in case the offsets topic somehow ends up with garbage in it, it'd be nice to allow users to clean it up via the REST API, and if there's no existing partition/offset pair in the offsets topic for the given partition already, then emitting a tombstone effectively becomes a no-op

Hm this is an interesting one. I think if the offsets topic ends up with garbage in it through an external producer, users could always clean it up using the same external producer. If we don't let users write garbage using the offsets REST API, maybe we shouldn't let them clean up garbage using the REST API either? Overall, I think I'm more in favor of only allowing standard use cases with the offsets management REST APIs to reduce the potential number of footguns for users.

Comment (C0urante) [reply connect/mirror/src/test/java/org/apache/kafka/connect/mirror/integration/MirrorConnectorsIntegrationExactlyOnceTest.java:88]: > Since the set of source partitions is limited here, shouldn't log compaction be good enough?

It's hard to define "limited" in this sense, but if you've configured MM2 to replicate every non-internal topic from a large cluster, then you could easily end up with hundreds of thousands of unique source offsets. And then, if some topics are deleted from the source cluster and other topics are created, we could go beyond even that. I think it'd be nice to allow people to do some cleanup in cases like that.

> If these offsets are never read back and actually used, why would users want to "undo" partial or complete resets?

I tried to touch on this with my earlier point:

> This is especially relevant since, although the offsets topic isn't public API, its contents are now public API (via the `GET /connectors/{name}/offsets endpoint`), and users may want to track the offsets for these connectors for monitoring purposes

TL;DR: The contents of the offsets topic may become an additional point of observability for the connectors for people to discover, e.g., the total set of replicated topic partitions or consumer groups.

> Overall, I think I'm more in favor of only allowing standard use cases with the offsets management REST APIs to reduce the potential number of footguns for users.

I definitely agree with the general mentality here of reducing the API surface in order to minimize the potential for users to hurt themselves. But is there any actual risk in the specific case of publishing tombstones for arbitrary topic partitions?

In general I'd still like it if we could provide a fairly flexible API for these connectors, but if that's too risky, one alternative could be to only permit tombstones with validated source partitions. This would still allow for cleanup (with no distinction between total and partial resets), but wouldn't support an "undo" for total/partial resets, or removal of garbage source partitions.

Comment (yashmayya) [reply connect/mirror/src/test/java/org/apache/kafka/connect/mirror/integration/MirrorConnectorsIntegrationExactlyOnceTest.java:88]: > It's hard to define "limited" in this sense, but if you've configured MM2 to replicate every non-internal topic from a large cluster, then you could easily end up with hundreds of thousands of unique source offsets. And then, if some topics are deleted from the source cluster and other topics are created, we could go beyond even that. I think it'd be nice to allow people to do some cleanup in cases like that.

Thanks for the explanation, I think that makes sense.

> TL;DR: The contents of the offsets topic may become an additional point of observability for the connectors for people to discover, e.g., the total set of replicated topic partitions or consumer groups.

Ah, my bad, I did not clock that as a potential reason for the undo partial reset functionality. I agree that if we're deciding to allow tombstoning offsets for partitions, we should also allow the reverse operation.

> I definitely agree with the general mentality here of reducing the API surface in order to minimize the potential for users to hurt themselves. But is there any actual risk in the specific case of publishing tombstones for arbitrary topic partitions?

I don't think there's any real "risk" as such; more of a general concern on the flexibility - user friendliness tradeoff. Although thinking about this a bit more, a new / unfamiliar user can't really do any harm by accidentally writing a tombstone for an arbitrary partition (and an error response is also probably not gonna be terribly useful) so I think I'm okay with abandoning source partition validation when the source offset value is `null`. As an aside, my unfamiliarity with running or operating MM2 is making me averse to bikeshedding too much over this issue (FWICT these scenarios should be rare enough that any of the options we've discussed should be acceptable), so I'll defer to your expertise here 😄

Comment (C0urante) [reply connect/mirror/src/test/java/org/apache/kafka/connect/mirror/integration/MirrorConnectorsIntegrationExactlyOnceTest.java:88]: 👍 Thanks Yash. I've pushed a change that allows arbitrary source partitions to be used with null source offsets; LMKWYT.

I'll leave this conversation unresolved in case the next reviewer wants to weigh in; your points are definitely worth considering!

Comment (yashmayya) [reply connect/mirror/src/test/java/org/apache/kafka/connect/mirror/integration/MirrorConnectorsIntegrationExactlyOnceTest.java:88]: > I've pushed a change that allows arbitrary source partitions to be used with null source offsets; LMKWYT.

Thanks Chris, looks good. We can probably update the `FileStreamSourceConnector::alterOffsets` method with a similar change in a follow up as well.

**Comment by C0urante** [comment]
@gharris1727 @mimaison would either of you be able to give this change a look? Hoping to get it merged in time for the 3.6.0 release if possible!

File: connect/mirror/src/test/java/org/apache/kafka/connect/mirror/integration/MirrorConnectorsIntegrationBaseTest.java [hunk 5]
```diff
@@ -933,12 +938,94 @@ protected static void waitUntilMirrorMakerIsRunning(EmbeddedConnectCluster conne
...
+            String connectorName = connectorClass.getSimpleName();
+            connectCluster.assertions().assertConnectorAndExactlyNumTasksAreRunning(
+                    connectorName,
+                    1,
+                    "Connector '" + connectorName + "' and/or task did not resume in time"
+            );
+        }
+    }
+
+    protected static void alterMirrorMakerSourceConnectorOffsets(EmbeddedConnectCluster connectCluster, LongUnaryOperator alterOffset, String... topics) {
```
Comment (gharris1727) [inline connect/mirror/src/test/java/org/apache/kafka/connect/mirror/integration/MirrorConnectorsIntegrationBaseTest.java:976]: This also stops the connectors, which isn't described in the function name. Can you split this into two functions, or change the name to stopAndAlter.. etc?

File: connect/mirror/src/test/java/org/apache/kafka/connect/mirror/integration/MirrorConnectorsIntegrationBaseTest.java [hunk 6]
```diff
@@ -933,12 +938,94 @@ protected static void waitUntilMirrorMakerIsRunning(EmbeddedConnectCluster conne
...
+        Set<String> topicsSet = new HashSet<>(Arrays.asList(topics));
+
+        String connectorName = MirrorSourceConnector.class.getSimpleName();
+        connectCluster.stopConnector(connectorName);
```
Comment (gharris1727) [inline connect/mirror/src/test/java/org/apache/kafka/connect/mirror/integration/MirrorConnectorsIntegrationBaseTest.java]: Should this additionally wait for the connectors to stop? it appears that it just waits for the REST API to return a 200, which may complete before the tasks have stopped executing and committed offsets.

File: connect/mirror/src/test/java/org/apache/kafka/connect/mirror/integration/MirrorConnectorsIntegrationExactlyOnceTest.java (hunk 4 above)
Comment (gharris1727) [reply connect/mirror/src/test/java/org/apache/kafka/connect/mirror/integration/MirrorConnectorsIntegrationExactlyOnceTest.java:88]: I think the currently proposed validation is reasonable, and I agree with the points raised above.
I didn't even know that the Checkpoint and Heartbeat connectors emitted offsets in the first place, since the operation of the connector never reads them back and I never had any use to trace their data flow. I think the GET portion of the API will be the first time any user sees these offsets, and the PUT / alterOffsets methods will be the first time that they see how they (don't) affect the operation of the connectors.

Since it is harmless to have the users change these offsets, the alterOffsets calls should guide users to well-formed inputs but not restrict them more than that. For the connector, the biggest impact will be keeping the offsets store clean in case a future extension wishes to use the offsets in a meaningful way.

**Review by gharris1727 — COMMENTED** [review]
Chris, thanks for the PR!

The implementation looks good, I only had a couple of concerns about testing.

File: connect/mirror/src/main/java/org/apache/kafka/connect/mirror/MirrorUtils.java [hunk 7]
```diff
@@ -65,27 +72,170 @@ static String encodeTopicPartition(TopicPartition topicPartition) {
...
+
+    /**
+     * Validate a specific key in a source partition that may be written to the offsets topic for one of the MM2 connectors.
+     * This method ensures that the key is present in the source partition map and that its value is a string.
+     *
+     * @see org.apache.kafka.connect.source.SourceConnector#alterOffsets(Map, Map)
+     * @see SourceRecord#sourcePartition()
+     *
+     * @param sourcePartition the to-be-validated source partition; may not be null
+     * @param key the key to check for in the source partition; may be null
```
Comment (mimaison) [inline connect/mirror/src/main/java/org/apache/kafka/connect/mirror/MirrorUtils.java]: It's not public API, all existing callers pass a non-null value and having null should cause an NPE below in get(), so maybe we can remove `may be null`?

File: connect/mirror/src/main/java/org/apache/kafka/connect/mirror/MirrorHeartbeatConnector.java (hunk 3 above)
Comment (mimaison) [reply connect/mirror/src/main/java/org/apache/kafka/connect/mirror/MirrorHeartbeatConnector.java:93]: Yeah I've not looked at the background but I agree with Yash, it looks strange

File: connect/mirror/src/main/java/org/apache/kafka/connect/mirror/MirrorUtils.java [hunk 8]
```diff
@@ -65,27 +72,170 @@ static String encodeTopicPartition(TopicPartition topicPartition) {
...
+     * Note that the partition key most likely refers to a partition in a Kafka topic, whereas the term "source partition" refers
+     * to a {@link SourceRecord#sourcePartition() source partition} that is stored in a Kafka Connect worker's internal offsets
+     * topic (or, if running in standalone mode, offsets file).
+     *
+     * @see org.apache.kafka.connect.source.SourceConnector#alterOffsets(Map, Map)
+     * @see SourceRecord#sourcePartition()
+     *
+     * @param sourcePartition the to-be-validated source partition; may not be null
+     *
+     * @throws ConnectException if the offset is invalid
```
Comment (mimaison) [inline connect/mirror/src/main/java/org/apache/kafka/connect/mirror/MirrorUtils.java]: nit: offset -> partition

File: connect/mirror/src/main/java/org/apache/kafka/connect/mirror/MirrorUtils.java [hunk 9]
```diff
@@ -65,27 +72,170 @@ static String encodeTopicPartition(TopicPartition topicPartition) {
...
+     * @param sourcePartition the corresponding {@link SourceRecord#sourcePartition() source partition} for the offset;
+     *                        may not be null
+     * @param sourceOffset the to-be-validated source offset; may be null (which is considered valid)
+     * @param permitPositiveValues whether positive values for the "offset" value in the source offset map
+     *                             should be permitted; if {@code true}, then all non-negative values are permitted; if
+     *                             {@code false}, only the value zero is permitted
+     *
+     * @throws ConnectException if the offset is invalid
+     */
+    static void validateSourceOffset(Map<String, ?> sourcePartition, Map<String, ?> sourceOffset, boolean permitPositiveValues) {
```
Comment (mimaison) [inline connect/mirror/src/main/java/org/apache/kafka/connect/mirror/MirrorUtils.java]: Would `onlyPartitionZero` be a clearer name?

**Review by mimaison — COMMENTED** [review]
Thanks for the PR, I left a few minor comments

File: connect/mirror/src/main/java/org/apache/kafka/connect/mirror/MirrorUtils.java (hunk 7 above)
Comment (C0urante) [reply connect/mirror/src/main/java/org/apache/kafka/connect/mirror/MirrorUtils.java]: It's unlikely that `Map::get` will lead to an NPE; it's optional for maps to refuse to allow null keys/values but neither `HashMap` nor `LinkedHashMap` have this restriction. But you're correct that no callers pass in null keys and I doubt we'll need support for that in the future, so will make the change 👍

File: connect/mirror/src/main/java/org/apache/kafka/connect/mirror/MirrorUtils.java (hunk 8 above)
Comment (C0urante) [reply connect/mirror/src/main/java/org/apache/kafka/connect/mirror/MirrorUtils.java]: Nice catch, thanks!

File: connect/mirror/src/main/java/org/apache/kafka/connect/mirror/MirrorUtils.java (hunk 9 above)
Comment (C0urante) [reply connect/mirror/src/main/java/org/apache/kafka/connect/mirror/MirrorUtils.java]: Maybe `onlyOffsetZero` since this deals with validation of the source offset instead of the source partition?

File: connect/mirror/src/test/java/org/apache/kafka/connect/mirror/integration/MirrorConnectorsIntegrationBaseTest.java (hunk 5 above)
Comment (C0urante) [reply connect/mirror/src/test/java/org/apache/kafka/connect/mirror/integration/MirrorConnectorsIntegrationBaseTest.java:976]: Ah yep, good call 👍 Split into two separate methods.

File: connect/runtime/src/test/java/org/apache/kafka/connect/util/clusters/EmbeddedKafkaCluster.java (hunk 1 above)
Comment (C0urante) [reply connect/runtime/src/test/java/org/apache/kafka/connect/util/clusters/EmbeddedKafkaCluster.java:600]: Did the same in `assertConnectorAndExactlyNumTasksAreRunning`.

File: connect/mirror/src/test/java/org/apache/kafka/connect/mirror/integration/MirrorConnectorsIntegrationBaseTest.java (hunk 6 above)
Comment (C0urante) [reply connect/mirror/src/test/java/org/apache/kafka/connect/mirror/integration/MirrorConnectorsIntegrationBaseTest.java]: Good catch! I've added a call to `connectCluster.assertions().assertConnectorIsStopped` whenever we stop a connector in these tests.

**Review by C0urante — COMMENTED** [review]
Thanks @gharris1727 and @mimaison! I've addressed your comments in the latest commit, except Mickael's question about the semantics of the various `alterOffsets` methods.

For reference, the full context is available [here](https://github.com/apache/kafka/pull/14005#discussion_r1262425850). Summarizing as best as I can (mostly copying from this [comment](https://github.com/apache/kafka/pull/14005#discussion_r1262913499)), the rationale is:

- Users may want to tombstone offsets for these connectors to prevent the offsets topic from growing infinitely
- We have no way of knowing if the set of offsets given to us is the total set or not, so we can't choose to only allow total resets instead of partial resets
- Since partial resets become possible, it seems reasonable to allow users to "undo" a partial reset by re-submitting an offset for a given partition
- This is especially relevant since, although the offsets topic isn't public API, its contents are now public API (via the `GET /connectors/{name}/offsets` endpoint), and users may want to track the offsets for these connectors for monitoring purposes
- We may make use of these offsets in future versions of these connectors, at which point, it'll be nice to know that the only offsets that we can ever encounter are the ones that were either emitted by the connector, or at least match the format of ones emitted by the connector
- This validation need not apply for tombstones, since emitting a tombstone for an unrecognized source partition is essentially a no-op and could be useful in case there's garbage in the offsets topic for a connector that the user would like to clean up

**Comment by C0urante** [comment]
@mimaison did you want to take another look at this? Curious about your take on the semantics for these connectors' `alterOffsets` methods, especially as an opportunity for us to set precedent for how other connector developers may implement that method.

**Comment by C0urante** [comment]
@gharris1727 do you think you could take a look? Feature freeze for 3.6.0 is Wednesday; was hoping to land this in time if possible.

**Review by gharris1727 — APPROVED** [review]
LGTM, Thanks @C0urante for adding this validation!

**Comment by C0urante** [comment]
Thanks @gharris1727!

@mimaison If you'd still like to weigh in, I can leave this open until Wednesday (two days from now).
